Introduce Colormaps

jonaspleyer opened the pull request:
This PR aims to add Colormaps to plotters. It is only a first version with almost no documentation.
I am happy to discuss the actual implementations and traits and write some documentation when the design is finalized.
To sum up my efforts:
- Define a trait for a colormap
    - Take a (generic) floating point value (between 0.0 and 1.0) and return a corresponding color
    - Can also be called by normalizing with a provided minimum and maximum value
- Define macros to create a colormap from colorvalues or existing colors by linearly interpolating between the individual colors
    - This trait defines individual structs which names are the names of the colormap (ie. `Viridis {}`)
    - These structs can be called by their implemented trait functionality after they have been instantiated
    - or by associated functions (which currently do not belong to the trait but this could be changed).
- I have added a new feature "colormaps" and added it to the default features to optionally deactivate this functionality.
  This is of course debateable and could also be seen as opt-in rather than opt-out.
- I have modified the examples for 3d-plot2 and mandelbrot to showcase the new colormaps
- In addition I have added a new example showcasing currently existing colormaps
- There are still many opportunities to define new and beautiful colormaps

Feel free to critique me in my approach. It was simply a feature I needed myself while using the library which I would like to see added. Thanks for writing this very nice piece of software!

AaronErhardt commented on plotters/src/style/colors/colormaps.rs:
```diff
@@ -0,0 +1,252 @@
+macro_rules! implement_linear_interpolation_color_map{
+    ($color_scale_name:ident, $color_type:tt) => {
+        impl<FloatType: std::fmt::Debug + Float + FromPrimitive + ToPrimitive> ColorMap<$color_type, FloatType> for $color_scale_name {
+            fn get_color_normalized(&self, h: FloatType, min: FloatType, max: FloatType) -> $color_type {
+                // Ensure that we do have a value in bounds
+                let h = h.max(min).min(max);
+                // Make sure that we really have a minimal value which is smaller than the maximal value
+                assert_eq!(min<max, true);
+                // Next calculate a normalized value between 0.0 and 1.0
+                let t = (h - min)/(max-min);
+                let approximate_index = t * (FloatType::from_usize(Self::COLORS.len()).unwrap() - FloatType::one()).max(FloatType::zero());
+                // Calculate which index are the two most nearest of the supplied value
+                let index_lower = approximate_index.floor().to_usize().unwrap();
+                let index_upper = approximate_index.ceil().to_usize().unwrap();
+                // Calculate the relative difference, ie. is the actual value more towards the color of index_upper or index_lower?
+                let relative_difference = approximate_index.ceil() - approximate_index;
+                // Interpolate the final color linearly
+                calculate_new_color_value!(relative_difference, Self::COLORS, index_upper, index_lower, $color_type)
+            }
```
There's a lot of duplicated code shared with `implement_color_scale_for_derived_ColorMap` that should probably be moved into a function.

jonaspleyer replied:
I have encapsulated the functionality into a function just as you described.

AaronErhardt commented on plotters/src/style/colors/colormaps.rs:
```diff
@@ -0,0 +1,252 @@
+
+
+macro_rules! implement_linear_interpolation_color_map{
+    ($color_scale_name:ident, $color_type:tt) => {
```
```suggestion
    ($color_scale_name:ident, $color_type:ty) => {
```

jonaspleyer replied:
See my comment above :smiley:

AaronErhardt commented on plotters/src/style/colors/colormaps.rs:
```diff
@@ -0,0 +1,252 @@
+);
+
+
+macro_rules! implement_color_scale_for_derived_ColorMap{
```
This macro doesn't follow the naming conventions and should be called `implement_color_scale_for_derived_color_map` instead.

jonaspleyer replied:
was renamed :+1:

AaronErhardt reviewed (changes requested):
+ Prefer functions over macros due to better readability and cleaner scope. Your code uses a lot of macros, which makes sense in most places, but where it is possible, functions should be used instead (e.g. in the new colormaps example).
+ Use cargo fmt to format the code consistently. You can use `#[rustfmt::skip]` for code you want to keep formatted manually.
+ Replace `tt` (token tree which could be anything) with `ty` (which can only be a type). I already commented on some occurrences, but this has been done wrong in most macros.

jonaspleyer commented:
Thank you for taking the time to review my code! I have a few deadlines coming up at the moment but hope to be getting back to you afterwards. That will probably be in about 1-2 weeks.

jonaspleyer commented on plotters/src/style/colors/colormaps.rs:
```diff
@@ -0,0 +1,252 @@
+
+
+macro_rules! implement_color_scale_for_derived_ColorMap{
+    ($($color_type:tt),+) => {
```
I have corrected it :+1: It does not appear to be working with `ty` but rather with `ident`. This is probably due to how I match the supplied patterns. eg these lines:
```
($relative_difference:expr, $colors:expr, $index_upper:expr, $index_lower:expr, HSLColor) => {
```
I could not manage to make it work by actually accepting a type instead of an identifier but am not sure why.

jonaspleyer commented on plotters/src/style/colors/colormaps.rs:
```diff
@@ -0,0 +1,252 @@
+            impl<FloatType: Float + FromPrimitive + ToPrimitive> ColorMap<$color_type, FloatType> for DerivedColorMap<$color_type> {
+                fn get_color_normalized(&self, h: FloatType, min: FloatType, max: FloatType) -> $color_type {
+                    // Ensure that we do have a value in bounds
+                    let h = h.max(min).min(max);
+                    // Make sure that we really have a minimal value which is smaller than the maximal value
+                    assert_eq!(min<max, true);
```
Yes this makes more sense. Looks like some debug assert statement that I forgot.

jonaspleyer commented:
I have addressed the concerns you had and have added documentation to the individual components. Please let me know if you still find that something is missing.

jonaspleyer commented:
Sorry for not noticing. Probably forgot to check again. Should be resolved now.

wangjiawen2013 commented:
@jonaspleyer 
Hi,
Could you present a tutorial on how to use colormap ?  Can I add a colorbar to a plot using colormap ?

jonaspleyer commented:
I have not implemented colorbars next to the plots. Currently their functionality is rather limited. Some of the examples already use said colormaps. Have a look eg. at the Mandelbrot example. It has been some time since I have dealt with this implementation so I would need to take a closer look in order to really decide how to best proceed on this topic. Unfortunately I have many other obligations at the moment.

Do you have any specific use-case in mind or is this a general question?

wangjiawen2013 commented:
colorbar is used frequently in my work. I add it to my plots to show the continuous data.

jonaspleyer commented:
I have to be honest: I cannot justify spending any time on this request currently. I might revisit this in the future however.
